server/conn_stmt.go

```diff
@@ -200,13 +199,18 @@ func (cc *clientConn) handleStmtExecute(ctx context.Context, data []byte) (err e
 	ctx = context.WithValue(ctx, execdetails.StmtExecDetailKey, &execdetails.StmtExecDetails{})
 	ctx = context.WithValue(ctx, util.ExecDetailsKey, &util.ExecDetails{})
 	retryable, err := cc.executePreparedStmtAndWriteResult(ctx, stmt, args, useCursor)
-	if retryable && err != nil && cc.ctx.GetSessionVars().IsRcCheckTsRetryable(err) {
-		logutil.Logger(ctx).Info("RC read using start_ts has failed, retry RC read",
-			zap.String("sql", cc.ctx.GetSessionVars().StmtCtx.OriginalSQL))
-		cc.ctx.GetSessionVars().RetryInfo.Retrying = true
-		_, err = cc.executePreparedStmtAndWriteResult(ctx, stmt, args, useCursor)
-		cc.ctx.GetSessionVars().RetryInfo.Retrying = false
-		return err
+	if err != nil {
+		action, txnErr := sessiontxn.GetTxnManager(&cc.ctx).OnStmtError(sessiontxn.StmtErrAfterQuery, err)
```

Comment:
The method name `OnStmtError` looks a bit strange to me here, by calling this method, we decide whether to retry this statement, and in line 210, there may also be a statement error, but the method `OnStmtError` is not called.

Reply:
`OnStmtError` does not require to hook every error occurs in the statement, its first parameter is `StmtErrorHandlePoint` so it only requires the error occurs in some specified points should be called.

 `CanStmtRetry` seems to be a method without any side effects and only gives a retry advice, but it is not, it also  changes the inner state of a provider.

Another concern is sometimes in some cases `OnStmtError` should also be called even if we should not retry. Because it should adjust some inner state for next statement.

Maybe we can give a longer name instead of `OnStmtError` to brief its function? For example `OnStmtErrorWithHandlePoint` ... I think retry is only one of the actions after the error, so I use a enum instead of bool as the return value.